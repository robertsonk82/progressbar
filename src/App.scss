.App {
  text-align: center;
}

.contentFooter{
  margin-top:8px;
  button{
    border:1px solid #999;
    margin:5px;
    border-radius:2px;
    background:#ccc;
  }
}