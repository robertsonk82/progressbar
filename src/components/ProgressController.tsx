
import React, { useState} from "react";
import ProgressBar from "./ProgressBar"

type ApiData = {
buttons: Array<number>;
bars: Array<number>;
limit: number
};

interface Props{
data: ApiData;
}

function ProgressController(props: Props) {
let { data } = props;

let [apiData, setApiData] = useState<ApiData>(data);

let [selectedProgressBar, setSelectedProgressBar] = useState(0);

function handleClick(event: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
    let buttonValue = event.currentTarget.dataset.value || "0";

    apiData.bars[selectedProgressBar] += parseInt(buttonValue);
    if (apiData.bars[selectedProgressBar] < 0) {
        apiData.bars[selectedProgressBar]= 0;
    }
    setApiData({...apiData});
};

function handleSelectChange(event: React.ChangeEvent<HTMLSelectElement>) {
    setSelectedProgressBar(parseInt(event.target.value));
}

//preparing the Progressbar View//
let progressView = apiData.bars.map((bar, i)=>(
<div key={i}>
    <ProgressBar progressValue={bar} maxLimit={apiData.limit || 0} />
</div>
));

let optionsView = apiData.bars.map((bar,i) =>(
    <option value={i} key={i}>
        Progress Bar {i+1}
    </option>
));

let selectView = (
    <select value={selectedProgressBar} onChange={handleSelectChange}>
        {optionsView}
    </select>
);

let buttonsView = apiData.buttons.map((button, i) =>(
    <button key={i} data-value={button} onClick={handleClick}>
        {button}
    </button>
));

return(
    <>
    {progressView}
    <div className="contentFooter">
        {selectView}
        {buttonsView}
    </div>
    </>
);
}
export default ProgressController;
