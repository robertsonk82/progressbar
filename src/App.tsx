import React, {useState, useEffect} from 'react';
import './App.scss';
import axios from "axios";
import ProgressController from "./components/ProgressController"



const App: React.FC = () => {
  let [apiData, setApiData]=useState();
  useEffect(()=>{
    axios.get("http://pb-api.herokuapp.com/bars").then(response =>setApiData(response.data))
  },[]);

  return (
    <div className="App">
      <h1>Pragress Bar Demo</h1>
      {apiData && <ProgressController data={apiData}/>}
    </div>
  )
}

export default App;
