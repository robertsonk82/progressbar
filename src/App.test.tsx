import React from 'react';
import {render, cleanup, fireEvent} from "@testing-library/react";
import "@testing-library/jest-dom";

import App from './App';
import ProgressController from "./components/ProgressController"

afterEach(cleanup);
let apiData = {"buttons":[12,15,-50,-48],"bars":[74,26,20],"limit":130}
it("matches snapshot", () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});

it("Render All Progress Elements", () => {
  const {container} = render(<ProgressController data={apiData} />);
  const buttons = container.querySelectorAll('button');
  const progressBars = container.querySelectorAll('.progress-text');
  const select = container.querySelector('select');
  const options = select!.querySelectorAll('option');
  expect(buttons.length).toBe(4);
  expect(progressBars.length).toBe(3);
  expect(options.length).toBe(3);
});

it("Fire button click event", () => {
  const { getByText} = render(<ProgressController data={apiData} />);
  const button = getByText("12");
  const progressBar1 = getByText("57%");

  fireEvent.click(button);
  expect(progressBar1).toHaveTextContent("66%")
});

it("Check value not be less than 0 ", () => {
  const { getByText} = render(<ProgressController data={apiData} />);
  const button = getByText("-50");
  const progressBar1 = getByText("20%");
  fireEvent.click(button);
  expect(progressBar1).toHaveTextContent("0%")
});

it("Select set 2nd progress bor and apply button click event", () => {
  const { container, getByText} = render(<ProgressController data={apiData} />);
  const button = getByText("12");
  const progressBar2 = getByText("20%");

  const select=container.querySelector("select");
  if(select){
    fireEvent.change(select,{target:{value:"1"}});
  }

  fireEvent.click(button);
  expect(progressBar2).toHaveTextContent("29%")
});

it("Check css color for the bar crossed above 100% ", () => {
  apiData = {"buttons":[70,15,-50,-48],"bars":[74,26,20],"limit":130}
  const {container, getByText} = render(<ProgressController data={apiData} />);
  const button = getByText("70");
  fireEvent.click(button);
  const progressBar1 = container.querySelectorAll('.crossedLimit');
  expect(progressBar1.length).toBe(1);
});
