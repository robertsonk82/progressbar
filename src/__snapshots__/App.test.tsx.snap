// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h1>
      Pragress Bar Demo
    </h1>
  </div>
</DocumentFragment>
`;
